torchci/scripts/check_alerts.py

```diff
@@ -0,0 +1,296 @@
+            else:
+                conclusions[SKIPPED] += 1
+        if conclusions[FAILURE] > 0 or conclusions[CANCELED]:
+            categorized_shas.append((sha, FAILURE))
+        elif conclusions[PENDING] > 0:
+            categorized_shas.append((sha, PENDING))
+        elif conclusions[SKIPPED] > 100:
+            categorized_shas.append((sha, SKIPPED))
+        else:
+            categorized_shas.append((sha, SUCCESS))
+    return categorized_shas
+
+
+def find_first_sha(categorized_sha: List[Tuple[str, str]], status: str):
+    for ind, sha in enumerate(categorized_sha):
+        if sha[1] == status:
+            return ind
+    return -1
+
+
+def clear_alerts(alerts: List[Any]) -> bool:
+    cleared_alerts = 0
+    for alert in alerts:
+        requests.patch(
+            UPDATE_ISSUE_URL + str(alert["number"]),
+            json={"state": "closed"},
+            headers=headers,
+        )
+        cleared_alerts += 1
+    print(f"Clearing {cleared_alerts} alerts.")
+    return cleared_alerts > 0
+
+
+# We need to clear alerts is there is a commit that's all green is before a commit that has a red
+# If there's pending things after the all green commit, that's fine, as long as it's all green/pending
+def should_clear_alerts(sha_grid: Any):
+    categorized_shas = categorize_shas(sha_grid)
+    first_green_sha_ind = find_first_sha(categorized_shas, SUCCESS)
+    first_red_sha_ind = find_first_sha(categorized_shas, FAILURE)
+    first_green = categorized_shas[first_green_sha_ind][0]
+    first_red = categorized_shas[first_red_sha_ind][0]
+
+    print(
+        f"The first green SHA was at index {first_green_sha_ind} at {first_green['sha']}"
+        + f"and the first red SHA was at index {first_red_sha_ind} at {first_red['sha']}"
+    )
+    if first_green_sha_ind < 0:
+        return False
+    return first_green_sha_ind < first_red_sha_ind
+
+
+# Creates Job Statuses which has the logic for if need to alert or if there's flaky jobs
+def classify_jobs(job_names: List[str], sha_grid: Any) -> Tuple[List[Any], List[Any]]:
+    job_data = map_job_data(job_names, sha_grid)
+    job_statuses: list[JobStatus] = []
+    for job in job_data:
+        job_statuses.append(JobStatus(job, job_data[job]))
+
+    jobs_to_alert_on = []
+    flaky_jobs = []
+
+    for job_status in job_statuses:
+        if job_status.should_alert():
+            jobs_to_alert_on.append(job_status)
+        flaky_jobs.extend(job_status.flaky_jobs)
+    return (jobs_to_alert_on, flaky_jobs)
+
+
+def main():
+    job_names, sha_grid = fetch_hud_data()
+    (jobs_to_alert_on, flaky_jobs) = classify_jobs(job_names, sha_grid)
+
+    # Fetch alerts
+    alerts = fetch_alerts()
+    alerts_cleared = False
+    # # Try to clear the alerts
+    if should_clear_alerts(sha_grid):
+        alerts_cleared = clear_alerts(alerts)
+
+    no_alert_currently_active = alerts_cleared == True or len(alerts) == 0
+    if len(jobs_to_alert_on) > 0 and (no_alert_currently_active):
```

Comment:
What happens if a second alert fires after the first one has been created?  It seems like it would skip creating an issue for the second alert

Reply:
Yea it would but at that point, the alert has already been fired and we don't need to alert again. I can maybe edit the already created alert to reflect the most up to date failures but I think I'll save that for a future PR.